### PR TITLE
Added simple Azure batch examples

### PR DESCRIPTION
**Pull request by lawrencegripper**
Hi, 

I'm currently doing some work using Azure Batch from Golang and thought it would be useful to pull it out as an example for others to use as I couldn't find any existing examples to follow. Would really welcome feedback. cc @alfpark 

This PR creates a simple example of how to create and then interact with an Azure Batch account through the Azure-sdk-for-go. 

To use the Azure Batch endpoints I've added an additional `GetBatchToken` call which gets a token scoped to azure batch's endpoint. I wasn't sure if this should be added to the `iam` package or directly in the `batch` package. Let me know your thoughts.

Ps. The work includes a workaround for the issue [here](https://github.com/Azure/azure-sdk-for-go/issues/1159) using a request inspector, hopefully this can be removed in the future. 

Example output:

```
~/go/src/github.com/Azure-Samples/azure-sdk-for-go-samples/batch$ go test 
2018/03/01 13:47:09 creating resource group 'azure-samples-go-CreateBatch-QkYkg' on location: westus2
2018/03/01 13:47:39 created batch account
2018/03/01 13:47:40 created batch pool
2018/03/01 13:47:41 created batch job
2018/03/01 13:47:41 created batch task
2018/03/01 13:50:44 output from task:
2018/03/01 13:50:44 Hello world from the Batch Hello world sample!

2018/03/01 13:50:44 deleting resources
PASS
ok  	github.com/Azure-Samples/azure-sdk-for-go-samples/batch	219.499s
```

### CODE REVIEW

File: iam/oauth.go [hunk 1]
```diff
@@ -120,6 +121,22 @@ func GetResourceManagementToken(grantType OAuthGrantType) (adal.OAuthTokenProvid
...
 
+const batchManagementEndpoint = "https://batch.core.windows.net/"
+
+// GetBatchToken gets an OAuth token for Azure batch using the specified grant type.
+func GetBatchToken(grantType OAuthGrantType) (adal.OAuthTokenProvider, error) {
+	if batchToken != nil {
+		return batchToken, nil
+	}
+
+	token, err := getToken(grantType, batchManagementEndpoint)
```
Comment (mcardosos) [inline iam/oauth.go]: Surprised this endpoint is not available on go-autorest. Creating issue

Comment (mcardosos) [reply iam/oauth.go]: https://github.com/Azure/go-autorest/issues/253

File: batch/batch.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+)
+
+const (
+	stdoutFile string = "stdout.txt"
+	stderrFile string = "stderr.txt"
+)
+
+// CreateAzureBatchAccount creates a new azure batch account
+func CreateAzureBatchAccount(ctx context.Context, accountName, location, resourceGroupName string) (a batchARM.Account, err error) {
+	token, err := iam.GetResourceManagementToken(iam.AuthGrantType())
```
Comment (mcardosos) [inline batch/batch.go]: On other samples we have preferred to move the client creation to its own function.

File: batch/batch.go [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+	accountClient := batchARM.NewAccountClient(helpers.SubscriptionID())
+	accountClient.Authorizer = autorest.NewBearerAuthorizer(token)
+	accountClient.AddToUserAgent(helpers.UserAgent())
+
+	res, err := accountClient.Create(ctx, resourceGroupName, accountName, batchARM.AccountCreateParameters{
+		Location: to.StringPtr(location),
+	})
+
+	if err != nil {
+		return batchARM.Account{}, err
```
Comment (mcardosos) [inline batch/batch.go]: This can be done like this `return a, err`

File: batch/batch.go [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+			},
+			NodeAgentSKUID: to.StringPtr("batch.node.ubuntu 16.04"),
+		},
+		MaxTasksPerNode:      to.Int32Ptr(1),
+		TargetDedicatedNodes: to.Int32Ptr(1),
+		// Create a startup task to run a script on each pool machine
+		StartTask: &batch.StartTask{
+			ResourceFiles: &[]batch.ResourceFile{
+				batch.ResourceFile{
+					BlobSource: to.StringPtr("https://gist.githubusercontent.com/lawrencegripper/795a9e809b52a0b8f874251c62a5a106/raw/8bfe4f4c80440204c994287e415c56f140dfd747/echohello.sh"),
```
Comment (mcardosos) [inline batch/batch.go]: Could this script live in this repo?

File: batch/batch.go [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+		VMSize: to.StringPtr("standard_a1"),
+	}
+
+	poolCreate, err := poolClient.Add(ctx, toCreate, nil, nil, nil, nil)
+
+	if err != nil {
+		return fmt.Errorf("cannot create pool: %v", err)
+	}
+
+	if poolCreate.StatusCode != 201 {
```
Comment (mcardosos) [inline batch/batch.go]: Is `200` status code not good?

File: batch/batch.go [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+	jobClient.AddToUserAgent(helpers.UserAgent())
+	jobClient.RequestInspector = fixContentTypeInspector()
+
+	jobToCreate := batch.JobAddParameter{
+		ID: to.StringPtr(jobID),
+		PoolInfo: &batch.PoolInformation{
+			PoolID: to.StringPtr(poolID),
+		},
+	}
+	// reqID := uuid.NewV4()
```
Comment (mcardosos) [inline batch/batch.go]: Please delete if not used

File: batch/batch.go [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+	res, err := jobClient.Add(ctx, jobToCreate, nil, nil, nil, nil)
+
+	if err != nil {
+		return err
+	}
+
+	if res.StatusCode != 201 {
```
Comment (mcardosos) [inline batch/batch.go]: What about `200`?

File: batch/batch.go [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+			},
+		},
+	}
+	res, err := taskClient.Add(ctx, jobID, taskToAdd, nil, nil, nil, nil)
+
+	if err != nil {
+		return "", err
+	}
+
+	if res.StatusCode != 201 {
```
Comment (mcardosos) [inline batch/batch.go]: 200?

File: batch/batch.go [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+// WaitForTaskResult polls the task and retreives it's stdout once it has completed
+func WaitForTaskResult(ctx context.Context, accountName, accountLocation, jobID, taskID string) (stdout string, err error) {
+	taskClient := must(getTaskClient(accountName, accountLocation))
+	res, err := taskClient.Get(ctx, jobID, taskID, "", "", nil, nil, nil, nil, "", "", nil, nil)
+	if err != nil {
+		return "", err
+	}
+
+	if res.State != batch.TaskStateCompleted {
+		for {
```
Comment (mcardosos) [inline batch/batch.go]: Could a time limit be added here? So it does not go on forever

File: batch/batch.go (hunk 4 above)
Comment (lawrencegripper) [reply batch/batch.go]: Sure, would make a lot of sense

File: batch/batch.go (hunk 5 above)
Comment (lawrencegripper) [reply batch/batch.go]: Probably being too pedantic - did it as the REST spec for Batch API says we're going to get a [201](https://docs.microsoft.com/en-us/rest/api/batchservice/pool/add) back but probably better to check for success code

Comment (lawrencegripper) [reply batch/batch.go]: Turns out it's all redundant anyway as autorest is returning err is it's not a 200 or 201 so I can remove these checks. https://github.com/Azure/azure-sdk-for-go/blob/master/services/batch/2017-09-01.6.0/batch/pool.go#L155

File: batch/batch_test.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+		accountName = strings.ToLower("b" + helpers.GetRandomLetterSequence(10))
+	}
+
+	jobID = strings.ToLower("j" + helpers.GetRandomLetterSequence(10))
+	poolID = strings.ToLower("p" + helpers.GetRandomLetterSequence(10))
+
+	clientID = os.Getenv("AZ_CLIENT_ID")
+	clientSecret = os.Getenv("AZ_CLIENT_SECRET")
+
+	return nil
```
Comment (mcardosos) [inline batch/batch_test.go]: What are this used for?

Comment (lawrencegripper) [reply batch/batch_test.go]: Looks like they sneaked in when copied the example from AKS sample. Will remove

**Pull request merged**